sdk/cosmos/azure-cosmos/azure/cosmos/_session_token_helpers.py

```diff
@@ -0,0 +1,206 @@
+# copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""Internal Helper functions for manipulating session tokens.
+"""
+from typing import Tuple, List
+
+from azure.cosmos._routing.routing_range import Range
+from azure.cosmos._vector_session_token import VectorSessionToken
+from ._feed_range import FeedRange
+
+# pylint: disable=protected-access
+
+def merge_session_tokens_with_same_range(session_token1: str, session_token2: str) -> str:
+    pk_range_id1, vector_session_token1 = parse_session_token(session_token1)
+    pk_range_id2, vector_session_token2 = parse_session_token(session_token2)
+    pk_range_id = pk_range_id1
+    if pk_range_id1 != pk_range_id2:
+        pk_range_id = pk_range_id1 \
+            if vector_session_token1.global_lsn > vector_session_token2.global_lsn else pk_range_id2
+    vector_session_token = vector_session_token1.merge(vector_session_token2)
+    return pk_range_id + ":" +  vector_session_token.session_token
+
+def is_compound_session_token(session_token: str) -> bool:
+    return "," in session_token
+
+def parse_session_token(session_token: str) -> Tuple[str, VectorSessionToken]:
+    tokens = session_token.split(":")
+    return tokens[0], VectorSessionToken.create(tokens[1])
+
+def split_compound_session_tokens(compound_session_tokens: List[Tuple[Range, str]]) -> List[str]:
+    session_tokens = []
+    for _, session_token in compound_session_tokens:
+        if is_compound_session_token(session_token):
+            tokens = session_token.split(",")
+            for token in tokens:
+                session_tokens.append(token)
+        else:
+            session_tokens.append(session_token)
+    return session_tokens
+
+def merge_session_tokens_for_same_partition(session_tokens: List[str]) -> List[str]:
+    i = 0
+    while i < len(session_tokens):
+        j = i + 1
+        while j < len(session_tokens):
+            pk_range_id1, vector_session_token1 = parse_session_token(session_tokens[i])
+            pk_range_id2, vector_session_token2 = parse_session_token(session_tokens[j])
+            if pk_range_id1 == pk_range_id2:
+                vector_session_token = vector_session_token1.merge(vector_session_token2)
+                session_tokens.append(pk_range_id1 + ":" + vector_session_token.session_token)
+                remove_session_tokens = [session_tokens[i], session_tokens[j]]
+                for token in remove_session_tokens:
+                    session_tokens.remove(token)
+                i = -1
```

Comment:
should it be` i--`?

Reply:
No I want to reset i to the beginning of the list because I want to include the merged session token I created because it could still be merged again with another session token and the indexing gets messed up after removing items from the list. Essentially:
1. Merge the session tokens with same pkrangeid 
2. then remove the old session tokens 
3. then reset to beginning 
repeat until no common pkrangeids. 
Alternative approach: 
1. Make lists of session tokens with common pkrangeids
2. Merge all the session tokens in those lists 
3. combine the lists

Reply:
Opinion: I would have done this taking extra space but favoring debuggability - in-place updates while also iterating over it seems like an anti-pattern.

Reply:
Also, in your example, good to have different `pkRangeId` in the list.

Reply:
Changed to the alternative